eq-author/src/components/QCodeContext/index.js

```diff
@@ -1,7 +1,92 @@
-import { createContext, useContext } from "react";
+import React, { createContext, useContext } from "react";
+import PropTypes from "prop-types";
+import CustomPropTypes from "custom-prop-types";
+
+import { RADIO } from "constants/answer-types";
 
 export const QCodeContext = createContext();
 
+export const getAllPages = ({ sections }) =>
+  sections?.flatMap(({ folders }) => folders?.flatMap(({ pages }) => pages));
+
+export const getFlattenedAnswerRows = (questionnaire) =>
+  getAllPages(questionnaire)
+    ?.filter(({ answers }) => answers?.length)
+    ?.flatMap(({ title, alias, answers }) =>
+      answers
+        .flatMap((answer) =>
+          [
+            answer,
+            ...(answer.options?.map((option) => ({
+              ...option,
+              type: answer.type === RADIO ? "RadioOption" : "CheckboxOption",
+              option: true,
+            })) ?? []),
+            answer.mutuallyExclusiveOption && {
+              ...answer.mutuallyExclusiveOption,
+              type: "MutuallyExclusiveOption",
+              option: true,
+            },
+            answer.secondaryLabel && {
+              id: answer.id,
+              label: answer.secondaryLabel,
+              qCode: answer.secondaryQCode,
+              type: answer.type,
+              validationErrorInfo: answer.validationErrorInfo,
+              secondary: true,
+            },
+          ].filter(Boolean)
+        )
+        .map((answer, index) => ({
+          ...answer,
+          questionTitle: title,
+          questionShortCode: alias,
+          nested: index > 0,
+        }))
+    );
+
+export const getDuplicatedQCodes = (flattenedAnswers) => {
+  const qCodeUsageMap = flattenedAnswers?.reduce((acc, item) => {
+    if (item.qCode) {
+      acc[item.qCode] = acc[item.qCode] ? acc[item.qCode] + 1 : 1;
+    }
+    return acc;
+  }, {});
+
+  // Just return a list of duplicated qCodes - counts are unimportant
+  return Object.entries(qCodeUsageMap)
+    .filter(([, count]) => count > 1)
+    .map(([qCode]) => qCode);
```

Comment:
```suggestion
  return Object.entries(qCodeUsageMap).reduce(
    (acc, [qCode, count]) => (count > 1 ? [...acc, qCode] : acc),
    []
  );
```
If you want to do it one time

Reply:
Done, in memoriam